iwyu_ast_util.cc

```diff
@@ -546,32 +562,30 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
-    return Base::TraverseType(type);
+    if (type.isNull())
+      return Base::TraverseType(type);
+    return TraverseTypeHelper(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
-    return Base::TraverseTypeLoc(type_loc);
-  }
-
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
+    if (!type_loc)
+      return Base::TraverseTypeLoc(type_loc);
+    return TraverseTypeHelper(type_loc.getType());
   }
 
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
-      return;
+  bool TraverseTypeHelper(QualType qual_type) {
+    CHECK_(!qual_type.isNull());
 
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+    const Type* desugared_until_alias_or_tpl = Desugar(qual_type.getTypePtr());
+    seen_types_.insert(desugared_until_alias_or_tpl);
+
+    const Type* desugared = DesugarAliasTypes(desugared_until_alias_or_tpl);
+
+    // Add desugared type components, if any. Don't add two types with the same
+    // canonical type.
```

Comment:
In fact, it doesn't always occur in this manner. Imagine `Template<A, TypedefForA>`. Maybe, it could be addressed in some future PRs...

Reply:
Ah, good catch. Please add a little TODO  so we remember it's a known weakness. It seems to me it should be solvable either by double-storing all seen canonical types, or by brute-force looping over `seen_types_` and comparing.

That small example would be nice to have as part of the TODO comment.

Reply:
Makes sense, of course. Done.